Assets/Scripts/NetworkClient.cs

```diff
@@ -78,30 +93,39 @@ private IEnumerator TryConnectToServers()
         {
             if (mainWebSocket == null)
             {
-                yield return new WaitForSeconds(1); // always wait 1s before next try
-
                 currentServerIndex++;
                 if (currentServerIndex >= _serverURLs.Count)
                 {
                     currentServerIndex = 0; // Reset to try again from the beginning.
                 }
 
                 string url = _serverURLs[currentServerIndex];
-                Debug.Log("Attempting to connect to: " + url);
+                Debug.Log("Attempting to connect to " + url);
 
                 var websocket = new WebSocket(url);
-                var connectTask = ConnectWebSocket(websocket, url);
+                FlagObject doAbort = new FlagObject();
+                // This is an async function that will finish later (but
+                // still on the main Unity thread).
+                var connectTask = ConnectWebSocket(websocket, url, doAbort);
+
+                // Wait for 4s, then check the result
+                yield return new WaitForSeconds(4);
 
-                // Wait for 8s, then cancel the connection attempt (i.e. time out)
-                yield return new WaitForSeconds(8);
+                if (mainWebSocket == null && websocket.State == WebSocketState.Connecting)
+                {
+                    // Wait another 4s
+                    yield return new WaitForSeconds(4);
+                }
 
                 if (mainWebSocket == null)
                 {
-                    Debug.LogWarning($"Timeout connecting to {url}.");
-                    // Note that this websocket object is in the middle of an async
-                    // call to Connect(), but this class appears safe to call
-                    // CancelConnection in this situation.
-                    websocket.CancelConnection();
```

Comment:
CancelConnection actually doesn't do anything useful for a socket still trying to connect. There is no way to cancel it. So I set a flag below and close it if/when it eventually gets opened.